### PR TITLE
LayerNorm test suite to fix launch kernels

### LINKED ISSUE #1376: Increase the test coverage for kernel launch failure with 6 arguments
Summary:
Tracks test coverage for PR #337 that fixed issue #54 (kernel launch failure with 6 arguments). This issue can be used to track all the testcase additions to increase the coverage.

Context:
As part of the PR #337, the kernel launch mechanism was fixed to properly handle kernels with 6 arguments (specifically for _layernormnorm_ operator). The fix modified the condition in torch_spyre/csrc/module.cp from _if (args.size() >= 3)_ to _if (args.size() == 6)_ and added proper handling for filling segment 5 to add another input to the device init supernode. This feature requires an extensive testing to uncover any scenarios missed to cover. A comprehensive test suite will help to identify any edge cases, ensure the fix works correctly  and capture any regressions in the future.

Coverage:
- [ ] PyTorch operators using 6-argument kernels (LayerNorm variants)
- [ ] Supported data types and tensor shapes
- [ ] Eager and compile execution modes
- [ ] Device and memory configurations
- [ ] negative scenarios

Acceptance Criteria:

- [ ] Define mandatory testcases and that should be passes 100% and enabled for eager/compile mode
- [ ] Accuracy validation against CPU
- [ ] Documentation of any failures with root cause analysis

### PR DESCRIPTION
**Pull request by neha-ibm**
This PR adds a comprehensive test suite to validate LayerNorm functionality and document backend limitations in the Torch-Spyre inductor backend.

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:
This PR adds a comprehensive test suite to validate LayerNorm functionality and document backend limitations in the Torch-Spyre inductor backend.

Following the fix in PR #337 for the 6-argument LayerNorm kernel, comprehensive test coverage is needed to:

1. **Prevent Regressions**: Guard against future breaks of the PR #337 fix
2. **Document Limitations**: Clearly document known backend behaviors and edge cases
3. **Validate Fixes**: Verify that all LayerNorm argument paths work correctly
4. **Guide Development**: Provide clear test cases for future improvements
5. **Issue Tracking**: Link tests to specific known issues for systematic resolution

**Test classes added:**

1. **TestLayerNormNegative** (8 tests)
2. **TestNormArgScaling** (11 tests)
3. **TestLayerNormBackendStress** (6 tests)
4. **TestLayerNormEdgeCases** (29 tests)

#### Which issue(s) this PR is related to:
**Issues this PR was raised to address:**

**Closes #1376** — Add comprehensive LayerNorm test suite. This PR directly addresses the request for systematic test coverage of LayerNorm functionality across all argument variations and edge cases.

**Related to PR #337** — LayerNorm 6-argument kernel fix. This test suite validates and guards the fix with specific regression tests for 3-6 argument paths.

**Related to Issue #54** — Original LayerNorm bug report that led to PR #337 and necessitated this comprehensive test coverage.

#### Issues documented during this testing work:

**Issue #1885 ** — C++ compilation errors in compiled mode. Tests document compilation failures for 4-5 argument LayerNorm calls with detailed xfail reasons.

**Issue #1888 ** — Numerical accuracy failures. Tests document 85-100% element mismatch vs CPU in both eager and compiled modes across multiple LayerNorm argument variations.

**Issue #1889** — Missing BatchNorm support. Test documents that `aten::native_batch_norm` is not implemented for Spyre backend, providing comparison point for normalization layer support.

**Issue #1813 - The Spyre backend's LayerNorm operation fails when processing non-contiguous tensors (views), such as those created by transpose(). 

#### Does this PR introduce a user-facing change?
No direct user-facing change. This PR adds internal test coverage.

#### Additional note:
**Platform: s390x | Python 3.12.12
**Test Results: 2 passed, 2 skipped, 99 xfailed, 5 xpassed in 51.25s

### CODE REVIEW

**Comment by dgrove-oss** [comment]
Feedback from coding agent review of this PR:

  ---
  Review

  BLOCKERs

  - Missing license header (tests/inductor/test_layernorm.py, line 1) — The file starts directly with import pytest. Every source file must carry the 14-line
  Apache 2.0 header. Compare with tests/inductor/test_inductor_ops.py:1–13.
  - Wrong test framework (tests/inductor/test_layernorm.py, entire file) — All other compiled-path tests in this repo use unittest.TestCase with the
  ParameterizedTestMeta metaclass (see test_inductor_ops.py:310). This PR uses plain pytest classes and @pytest.mark.parametrize. That diverges from the
  established pattern and won't integrate cleanly with the harness.
  - Double-underscore typo in method names — test_large__values (line 85), test_extreme__eps (line 826), test_zero__variance (line 840) all have double
  underscores. This looks unintentional and is inconsistent with the rest of the file.
  - torch.manual_seed called at class body level (TestLayerNormBackendStress:394, TestLayerNormEdgeCases:495) — These calls execute at import time, not per-test,
  so they don't seed anything reliably. The other test classes use setup_method for this. TestLayerNormNegative does it correctly; the other two classes should
  follow suit.
  - unlisted_test_mode: mandatory_success (tests/configs/torch_spyre_tests/inductor/test_layernorm_config.yaml:4) — With 99 xfailed and 1 skipped test (per the
  author's own results), this mode requires all non-xfail/skip tests to pass. This is appropriate only if the two passing tests (test_rmsnorm and presumably one
  other) are the only non-marked tests. Verify the config mode is intentional and correct given the test count.

  SUGGESTIONs

  - xfail issue references are vague — Several tests reference "Issue #1", "Issue #2", "Issue #3" (e.g. line 117, 135, 194). These should use real issue numbers
  (like #1885, #1888, #1889 mentioned in the PR description) so readers can follow up without reading the PR body.
  - test_batchnorm_safe placement — A BatchNorm test in a LayerNorm file is confusing. It makes sense as a comparison point, but consider a brief comment
  explaining the intent (e.g. "included as normalization coverage baseline").
  - PR title misleading — "Fix launch kernels" does not describe what this PR does (adds a LayerNorm test suite). The description is accurate; the title should
  match.

**Comment by dgrove-oss** [comment]
I also asked for an analysis of overlap of the new tests with existing test suite. 

Please consider this and decide whether it is worth pursuing this PR or if we should decide that the core issue is already covered with test cases that have been added in other PRs since this work began.

------

  The new PR's tests are largely redundant, and where they're novel, the novelty is mostly negative.

  What already exists

  test_inductor_ops.py (test_layernorm_cpu): Covers the standard F.layer_norm call with weight and bias in 2D, and also a transposed (non-contiguous) 2D input —
  using the ParameterizedTestMeta framework with stick-aligned shapes.

  test_inductor_normalization_scalars.py: A thorough normalization test file already covering:
  - Manual mean/var LayerNorm decompositions (2D, 3D, BERT-like)
  - LayerNorm with affine (gamma/beta)
  - RMSNorm with many shape/eps combinations
  - GroupNorm, InstanceNorm (1D, 2D, 3D), BatchNorm — all with xfail markers linked to real issue numbers

  What's new in PR #1895

  Very little that isn't already covered:
  - F.layer_norm via the nn.functional API (vs. manual decomposition) — this is a genuine gap, since the existing normalization file uses explicit mean/var
  decompositions rather than the ATen layer_norm kernel directly. test_inductor_ops.py does call F.layer_norm, but only for the 6-arg case with one shape.
  - F.batch_norm and F.group_norm via high-level API — the existing file tests these as manual decompositions, not via their ATen kernels. Same argument applies.

  The core problem

  Almost all 54 new tests are xfail. The existing file already documents the same failures (numerical inaccuracy, var/mean not implemented) with proper issue
  numbers. The PR adds ~900 lines to document that LayerNorm is broken in many ways — but that's already known and tracked. It doesn't add passing tests or cover
  a working code path that was previously untested.

  Bottom line: The PR adds marginal coverage of F.layer_norm via the functional API (which does exercise the ATen kernel path, not the decomposition), but the
  vast majority of tests are xfailed duplicates of known failures already documented in test_inductor_normalization_scalars.py.

**Comment by dgrove-oss** [comment]
Hi @neha-ibm  -- per the previous comment there is significant overlap with existing tests.  Do you want to pursue this PR by narrowing it or should we close it?  We should make a decision so we can bring the epic to a close.  Thanks.

**Comment by neha-ibm** [comment]
@dgrove-oss  I shall narrow it down to only the valid cases.

**Comment by neha-ibm** [comment]
Hi @dgrove-oss . I have narrowed down the test cases.

File: tests/configs/torch_spyre_tests/inductor/test_inductor_ops_misc_compute_b_config.yaml [hunk 1]
```diff
@@ -18,7 +18,13 @@ test_suite_config:
         - names:
             # normalization
             - TestOps::test_layernorm.*
+            - TestOps::test_layernorm_functional.*
             - TestOps::test_rmsnorm.*
+            - TestOps::test_rmsnorm_manual.*
+            - TestOps::test_batch_norm_functional.*
+            - TestOps::test_group_norm_functional.*
+            - TestOps::test_softmax.*
```
Comment (dgrove-oss) [inline tests/configs/torch_spyre_tests/inductor/test_inductor_ops_misc_compute_b_config.yaml:26]: Small nit.
The test_softmax.* and test_softplus.* lines here are duplications.  My guess is a merge conflict resolution that didn't quite do what you wanted.

**Review by dgrove-oss — COMMENTED** [review]
Thank you @neha-ibm.

One merge artifact to clean up in the test config yaml and then this can be merged.

File: tests/configs/torch_spyre_tests/inductor/test_inductor_ops_misc_compute_b_config.yaml (hunk 1 above)
Comment (neha-ibm) [reply tests/configs/torch_spyre_tests/inductor/test_inductor_ops_misc_compute_b_config.yaml:26]: Hi @dgrove-oss . Resolved this one.

**Review by KanakaPathivada — APPROVED** [review]

**Review by dgrove-oss — APPROVED** [review]
lgtm
